### ISSUE TITLE
Unnecessary CPU to GPU copies within flex_attention

### ISSUE THREAD

**Issue by VivekPanyam** [pytorch/pytorch#143927]
There are a few unnecessary CPU to GPU copies within flex_attention that cause unnecessary `cudaStreamSynchronize`s to occur. This decreases GPU utilization.

## Note

The below is based on a profiling run without `torch.compile`. I haven't looked at profiles of the compiled version in depth yet, but based on a quick glance, the second issue with `index_put_` seems to go away when compiled.

(My compiled tests were run with a patch for the first issue below so the results don't confirm whether the patch is necessary or if `compile` fixes it)

## Example 1

This one happens during every execution of attention:
https://github.com/pytorch/pytorch/blob/fe398de769480ceb943d3ae37551a29954bbb147/torch/_higher_order_ops/flex_attention.py#L189

That line creates a `-inf` tensor on CPU and copies it to GPU (along with a synchronize). It should be replaced with

```py
torch.full((), -float("inf"), dtype=working_precision, device=scores.device),
```

I put up a PR to fix it: #143928

## Example 2

In the following code, several `cudaStreamSynchronize`s happen within `aten::index_put_` (the `dense_mask[row_indices, valid_indices] = 1 ` line).

https://github.com/pytorch/pytorch/blob/fe398de769480ceb943d3ae37551a29954bbb147/torch/nn/attention/flex_attention.py#L152-L166

My profiling shows 3 `cudaStreamSynchronize`s per `create_dense_one` call (which is called multiple times in a single `create_block_mask` call). I was able to decrease it to 2 synchronizes by replacing

https://github.com/pytorch/pytorch/blob/fe398de769480ceb943d3ae37551a29954bbb147/torch/nn/attention/flex_attention.py#L165

with 

```py
dense_mask[row_indices, valid_indices] = torch.ones((), dtype=torch.int32, device=device)
```

The other two synchronizes are also within `aten::index_put_`, but they didn't jump out to me based on a cursory scan of the code.

cc @zou3519 @bdhirsh @penguinwu @yf225 @Chillee @drisspg @yanboliang @BoyuanFeng @chauhang @ydwu4

**Comment by vadimkantorov** [comment]
> That line creates a `-inf` tensor on CPU and copies it to GPU (along with a synchronize). It should be replaced with
> 
> ```python
> torch.full((), -float("inf"), dtype=working_precision, device=scores.device),
> ```

wow, this is super-surprising that torch.full versus torch.tensor (which is more frequently used in practice for this end IMO) makes such a difference for creating scalars. I wonder if torch.tensor itself should use the same fast sync-avoiding path when called with a python scalar?

**Comment by drisspg** [comment]
Hey so we haven't paid much attention to flex-attentions eager perf because it should be muchhhh slower than the alternative (scaled_dot_product_attention) when not using torch.compile. 

The second example in create-block-mask is somewhat similiar since would like users to also compile this function however I do have a PR to remove the sync: https://github.com/pytorch/pytorch/pull/143872

**Comment by Chillee** [comment]
We should have fixed the `create_block_mask` cuda syncs (which also wouldn't show if you compile it :P)
